doc/neps/nep-0028-website-redesign.rst

```diff
@@ -0,0 +1,302 @@
+welcoming.
+
+On the other hand, there are good reasons why few projects have a
+multi-lingual site. It's potentially a lot of extra work. Extra work for
+maintainers is costly - they're already struggling to keep up with the work
+load. Therefore we have to very carefully consider whether a multi-lingual
+site is feasible and weight costs and benefits.
+
+We start with an assertion: maintaining translations of all documentation, or
+even the whole user guide, as part of the NumPy project is not feasible. One
+simply has to look at the volume of our documentation and the frequency with
+which we change it to realize that that's the case. Perhaps it will be
+feasible though to translate just the top-level pages of the website. Those
+do not change very often, and it will be a limited amount of content (order
+of magnitude 5-10 pages of text).
+
+We propose the following requirements for adding a language:
+
+- The language must have a dedicated maintainer
+- There must be a way to validate content changes (e.g. a second
+  maintainer/reviewer, or high quality language support in a freely
+  available machine translation tool)
+- The language must have a reasonable size target audience (to be
+  assessed by the NumPy maintainers)
+
+Benefits include:
+
+- Better serve many existing and potential users
+- Potentially attract a culturally and geographically more diverse set of contributors
+
+Tradeoffs:
+
+- Cost of maintaining a more complex code base
+- Cost of making decisions about whether or not to add a new language
+- Higher cost to making content changes, creates work for language maintainers
+- Any content change should be rolled out with enough delay to have translations in place
+
+Can we define a small enough set of pages and content that it makes sense to do this?
+Probably yes.
+
+Is there an easy to use tool to maintain translations and add them to the website?
+To be discussed - it needs investigating, and may depend on the choice of static site
+generator. One potential option is Crowdin_, which is free for open source projects.
+
+
+Backward compatibility
+----------------------
+
+Given a static site generator is chosen, we will migrate away from Sphinx for
+numpy.org (the website, *not including the docs*). The current deployment can
+be preserved until a future deprecation date is decided (potentially based on
+the comfort level of our new site).
+
+All site generators listed above have visibility into the HTML and Javascript
+that is generated, and can continue to be maintained in the event a given
+project ceases to be maintained.
+
+
+Alternatives
+------------
+
+Alternatives we considered:
+
+1. *Update current site.* A new Sphinx theme could be chosen. This would likely
+   take the least amount of resources initially, however, Sphinx does not have
+   the features we are looking for moving forward such as i18n, responsive design,
+   and a clean, modern look.
```

Comment:
I disagree with this point. While I'm surely not a fan of Sphinx, it can have a clean modern look, and it has i18n support (not sure if there are limitations to it, but afaik content can be translated with `.po` files).

For pandas we're planning to do the opposite as this NEP proposes. We have https://pandas.pydata.org independent from the docs (it's also Sphinx, but could be any static website generator), and we're planning to integrate it with the docs, which should make things simpler to maintain.

We've got a prototype to experiment with the styles, you can see it here: https://dev.pandas.io/pandas-sphinx-theme/pr-datapythonista_base/. It's using https://github.com/ryan-roemer/sphinx-bootstrap-theme, the maintainers of the project have been helpful in the past, and they are open to adapt the theme for our needs.

It is responsive, and I would say the look is modern (at least compared to our current docs). If it's not, it can be improved with a designer modifying the css (I don't have great designing skills myself). I was experimenting on not having a navigation bar (that's why in the black header there are no links), but they come by default and we'll have them.

I'd personally be very happy to not have to deal with Sphinx, but since the plan is using it for the docs anyway, I think doesn't add much value to have a separate website, and two systems to maintain. Customizing Sphinx will require some work, but since we're going to do it for pandas, there will be no extra work to do for numpy/scipy.

And personally, I think the closer the docs/website of numpy, pandas and the rest of the ecosystem are, the better. Whatever is the final solution, I think we should follow the same approach, and also standardize the names of the sections and the main structure of the website/docs.

Reply:
> I disagree with this point. While I'm surely not a fan of Sphinx, it can have a clean modern look, and it has i18n support (not sure if there are limitations to it, but afaik content can be translated with `.po` files).

Thanks for pointing it out. I came across talk about [Multilingual support for Sphinx](https://www.slideshare.net/streetcleaner/writing-multi-language-documentation-using-sphinx).  I will explore it.

The idea behind this is - We will write on simple english and out Statis Site generator will translate it to multi languages like the way docusaurus do : https://docusaurus.io/docs/en/translation


> For pandas we're planning to do the opposite as this NEP proposes. We have https://pandas.pydata.org independent from the docs (it's also Sphinx, but could be any static website generator), and we're planning to integrate it with the docs, which should make things simpler to maintain.
> 
> We've got a prototype to experiment with the styles, you can see it here: https://dev.pandas.io/pandas-sphinx-theme/pr-datapythonista_base/. It's using https://github.com/ryan-roemer/sphinx-bootstrap-theme, the maintainers of the project have been helpful in the past, and they are open to adapt the theme for our needs.

Thanks for sharing these links.

> It is responsive, and I would say the look is modern (at least compared to our current docs). If it's not, it can be improved with a designer modifying the css (I don't have great designing skills myself). I was experimenting on not having a navigation bar (that's why in the black header there are no links), but they come by default and we'll have them.
> 
> I'd personally be very happy to not have to deal with Sphinx, but since the plan is using it for the docs anyway, I think doesn't add much value to have a separate website, and two systems to maintain. Customizing Sphinx will require some work, but since we're going to do it for pandas, there will be no extra work to do for numpy/scipy.

Actually, the plan is having one website (numpy.org) landing page (and all other static pages like About, User Guide, News, tutorials ) will be generated using SSG (docusaurus/jekyll/Hugo) and `numpy.org/docs` will contain the generated build of Sphinx (which will be having Python API docstrings).

So it may engage reader/contributors to modify the Static file because it will be written in simple Markdown (like plain text), so need to setup SSG  framework. Also writing blogs or news about numpy will be easy using SSG.
 
> And personally, I think the closer the docs/website of numpy, pandas and the rest of the ecosystem are, the better. Whatever is the final solution, I think we should follow the same approach, and also standardize the names of the sections and the main structure of the website/docs.

Reply:
> So it may engage reader/contributors to modify the Static file because it will be written in simple Markdown

I agree on this, but is this different than with Sphinx? If you're going to have the documentation in markdown for sphinx, having to maintain another system that does the same is what I don't understand. Or do you mean having markdown as opposed as Sphinx rst?

Reply:
> we're planning to integrate it with the docs, which should make things simpler to maintain.

This is a no-go I think. The website builds in seconds, the docs take 10 minutes and require things like LaTeX. We even have a separate build for NEPs (which is fine in Sphinx), because people felt it was better to have that build fast.

> While I'm surely not a fan of Sphinx, it can have a clean modern look

Everything is possible, but there are no examples of this I can think of (after a decade of all of us doing everything in Sphinx ....). Sphinx is for docs, not for websites. In theory you can build something like https://jupyter.org/ (or https://pytorch.org/ or https://www.tensorflow.org/) on Sphinx, but it will be extremely painful and much harder to maintain than Hugo or Jekyll.

I want to engage people with good web dev skills, rather than adapt the tool to the experience of the current maintainers. There won't be many web developers who will think it's reasonable to build a static site with smaller amounts of text but modern styling, some graphics, etc. in Sphinx.

> We've got a prototype to experiment with the styles, you can see it here: https://dev.pandas.io/pandas-sphinx-theme/pr-datapythonista_base/. 

That looks like a significant improvement over the current site, nice.

> And personally, I think the closer the docs/website of numpy, pandas and the rest of the ecosystem are, the better.

That may be true. I'd rather follow Jupyter though, as the current best site in our ecosystem.

Reply:
> I agree on this, but is this different than with Sphinx? If you're going to have the documentation in markdown for sphinx, having to maintain another system that does the same is what I don't understand. Or do you mean having markdown as opposed as Sphinx rst?

no we won't change the docs system - 1500 pages of text can't be changed away from Sphinx. 

it's just about numpy.org, which right now has about 1.5-2 pages worth of text.

Reply:
Just two comments on the above:

I think having a fresh, good looking website depends almost exclusively of the work in the CSS. If you have someone with good design skills, who spend enough time making the page look good, it will look good no matter if it's built with Sphinx, Pelican, Jekyll or whatever. Of course the html has to be generated in the appropriate way. There are limitations in every static site generator, and in Sphinx the amount of work to change things is higher (you can see https://github.com/pandas-dev/pandas/issues/23763 for reference of my feelings when I tried to customize the table of contents).

For me the trade off is between how annoying is to deal with Sphinx to customize the html, and how annoying is to maintain two separate systems.

The good news is that the CSS is somehow independent of how the html is generated. So, I don't think it's a big deal to make the wrong choice to start with. The amount of work to move from Sphinx to Jekyll or the opposite should be much smaller than the work to create the CSS I think.

Regarding the build time, in pandas we have a script `doc/make.py` that we can call with a `--single` parameter to generate a single page. We also have a parameter `--no-api` to skip the API documentation, which takes a most of the time to build. It'd also be easy to add a parameter to build only the website and not the docs, or to skip running the code examples, and Sphinx would be extremely fast.

One thing to consider is how often you want to update the website compared to the the docs. Meaning that if you have all together, and you want to change the home page, you need to wait for the next release for users to see it. We're considering having a blog built with the docs in Sphinx, and that would be a problem. There is a solution if your website is hosted behind nginx, which is to redirect to the dev docs with nginx rules (which would be transparent for the users, they'd still see the same domain). That is quite simple to implement and would be useful also for parts of the documentation, like the contributing guide, which users usually care about the most updated version and not the last stable.

So, my summary:
- Dealing with Sphinx is not great, but since you're going to use it for the docs, it's worth considering it for the website too
- I don't think how nice the website looks depends much on what you use to generate the html (more on the amount of work you put in the css, and sphinx has a setting to specify your custom css files, that's trivial)
- Having the website separate has surely some advantages, but I don't think they are the building time, how good the website looks, or being responsive. Probably they are not making people who deal with the website have to clone the whole numpy, and how often you can release
- pandas website is most of the time outdated, while the docs are updated, besides having to maintain two systems, deal with two repos, deploy two separate sites... I think that's the main reason to consider building the website with sphinx. Also having a single domain with unified style (you can address that with separate websites, but it's tricker).

CC: @jorisvandenbossche stijnvanhoey may be you are interested in this discussion

Reply:
thanks @datapythonista. that issue with Sphinx customization feels familiar:)

> Meaning that if you have all together, and you want to change the home page, you need to wait for the next release for users to see it.

yeah, I don't think a 1-6 month wait is acceptable. also, the website is conceptually independent from a release.

> Dealing with Sphinx is not great, but since you're going to use it for the docs, it's worth considering it for the website too

I'm having in mind that these are largely separate (groups of) people developing it. And in the end both are just a `make html`. 

> I don't think how nice the website looks depends much on what you use to generate the html

this is the point I disagree with most. experience shows that it's just not true for Sphinx. it may be true _in theory_, but that's like saying it doesn't matter whether you choose Perl or Python 

> pandas website is most of the time outdated,

the NumPy one too, and that's largely because people who care about code first and everything  else second. we now have a good amount of energy from new people (with better web dev skills, or interest in educational materials, marketing, etc.).

Reply:
For what it worth, we have a separate repo for the astropy webpage, and it is indeed very nice not to be tied to the release cycle of the repo (including not to have a different webpage for LTS/bugfix/latest releases).

Reply:
I feel the need to respond to this thread since I wrote what we're discussing here in the NEP :) 
I'll admit I took my own experiences with Sphinx into account without knowing much about its theming. I'll also admit I don't like Sphinx much, but it sounds like I'm not alone! At best, what I wrote in the NEP isn't very objective, so I'm open to editing this line.

I'd also like to preface I've been a front-end developer for years, so I'm hoping I can help steer a bit and correct any misconceptions, answer questions, and whatnot.

> For me the trade off is between how annoying is to deal with Sphinx to customize the html, and how annoying is to maintain two separate systems.

If it truly is _annoying_ to customize html in Sphinx, I think that alone is a dealbreaker for writing the landing page in Sphinx. You also mentioned:

> Probably they are not making people who deal with the website have to clone the whole numpy, and how often you can release

I haven't cloned NumPy, but it already sounds large. I'd imagine this would slow down development for a relatively small landing page, and beyond that, I doubt we'd want releases focused around landing page changes.

> Having the website separate has surely some advantages, but I don't think they are the building time, how good the website looks, or being responsive.

Build time is probably an advantage. I don't know how long it takes NumPy to build, but it sounds like it might take a little bit. Without NumPy (i.e. standalone Jekyll static page), build time should be seconds. You're right about the other two.

> I don't think how nice the website looks depends much on what you use to generate the html (more on the amount of work you put in the css, and sphinx has a setting to specify your custom css files, that's trivial)

You're right :) However, you mentioned:

> The good news is that the CSS is somehow independent of how the html is generated. So, I don't think it's a big deal to make the wrong choice to start with. The amount of work to move from Sphinx to Jekyll or the opposite should be much smaller than the work to create the CSS I think.

This unfortunately isn't accurate. I wish it was. CSS rules work differently depending on your HTML structure since CSS _does_ care about parent/child elements. So given some really decent CSS, we couldn't port _just_ the CSS without an identical HTML structure and expect it to work properly.

I think the docs and the website should be separated. However, it seems like there's consensus around the docs staying put in Sphinx for several packages (Pandas, AstroPy, etc.), which is great. I do think it would be a bonus if multiple packages _did_ have the same styling, which would make navigating much easier for users of all packages.

Reply:
We mostly agree, just two things I didn't explain well and I think you misunderstood.

It's very easy to have the website in the docs, and be able to call something like `doc/make.py --website-only` and the website build will take couple of seconds. In pandas we're doing similar things.

Editting the CSS in Sphinx is trivial (same as in whatever else). The reason is more complex to add style is because changing the html is not as easy as just changing the template for things like the toc tree. That's why I said moving from Sphinx to Jekyll or the other way is doable; the work needed would be just to make the html the same (or close enough for the css).

Also, worth mentioning that for most parts of the website, changing the Sphinx templates should be enough, and for special pages (like the home in the pandas new look I sent) html can be used directly.

Reply:
Understood, thank you for clarifying!